janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMultiQueryStrategy.java

```diff
@@ -0,0 +1,121 @@
+            || TraversalHelper.onGraphComputer(traversal)
+            // The LazyBarrierStrategy is not allowed to run on traversals which use drop(). As a precaution,
+            // this strategy should not run on those traversals either, because it can also insert barrier().
+            || !TraversalHelper.getStepsOfAssignableClassRecursively(DropStep.class, traversal).isEmpty()) {
+            return;
+        }
+
+        final StandardJanusGraph janusGraph = JanusGraphTraversalUtil.getJanusGraph(traversal);
+        if (janusGraph == null || !janusGraph.getConfiguration().useMultiQuery()) {
+            return;
+        }
+
+        insertMultiQuerySteps(traversal, janusGraph.getConfiguration().limitBatchSize());
+        configureMultiQueriables(traversal);
+    }
+
+    /**
+     * Insert JanusGraphMultiQuerySteps everywhere in the current traversal where MultiQueriable steps could benefit
+     *
+     * @param traversal The local traversal layer.
+     */
+    private void insertMultiQuerySteps(final Admin<?, ?> traversal, boolean limitBatchSize) {
+        JanusGraphTraversalUtil.getSteps(JanusGraphTraversalUtil::isMultiQueryCompatibleStep, traversal).forEach(step -> {
+            Optional<Step> multiQueryPosition = JanusGraphTraversalUtil.getLocalMultiQueryPositionForStep(step);
+            if (multiQueryPosition.isPresent() && JanusGraphTraversalUtil.isLegalMultiQueryPosition(multiQueryPosition.get())) {
+                Step pos = multiQueryPosition.get();
+                if (limitBatchSize && !(multiQueryPosition.get() instanceof NoOpBarrierStep)) {
+                    NoOpBarrierStep barrier = new NoOpBarrierStep(traversal);
+                    TraversalHelper.insertBeforeStep(barrier, pos, traversal);
+                    pos = barrier;
+                }
+                JanusGraphMultiQueryStep multiQueryStep = new JanusGraphMultiQueryStep(traversal, limitBatchSize);
+                TraversalHelper.insertBeforeStep(multiQueryStep, pos, traversal);
+            }
+        });
+    }
+
+    /**
+     * Looks for MultiQueriables in within the traversal and registers them as clients of their respective
+     * JanusGraphMultiQuerySteps
+     *
+     * @param traversal The local traversal layer.
+     */
+    private void configureMultiQueriables(final Admin<?, ?> traversal) {
+        TraversalHelper.getStepsOfAssignableClass(MultiQueriable.class, traversal).forEach(multiQueriable -> {
+            final List<Step> mqPositions = JanusGraphTraversalUtil.getAllMultiQueryPositionsForMultiQueriable(multiQueriable);
+            final List<JanusGraphMultiQueryStep> multiQuerySteps = new ArrayList<>(mqPositions.size());
+            boolean multiQueryAllowed = true;
+            for (Step mqPos : mqPositions) {
+                // If one position is not legal, this means that the entire step can not use the multiQuery feature.
+                multiQueryAllowed &= JanusGraphTraversalUtil.isLegalMultiQueryPosition(mqPos);
```

Comment:
nitpick: it seems if `multiQueryAllowed` becomes false, we can `break` or `return` right away.

Reply:
Yes, you're right. It is also possible to perform all checks first and then enter a second loop which does the work only if the checks have passed. I refactored this method and I think it's even easier to read now.